Replace slow dist_math::incomplete_beta with aesara op

ricardoV94 opened the pull request:
This PR replaces the `incomplete_beta` method in dist_math by an aesara op `betainc` that wraps the `scipy.special.betainc` method.

This provides a large increase in speed (and graph compilation) even without providing the c_code implementation directly (this can always be added later). It also allows the evaluation of the logcdf for multiple values in `Beta`, `StudentT`, `Binomial`, `NegativeBinomial` (and zero inflated variants). I removed the previous limitation warning when trying to evaluate these methods for multiple values. These changes are critical for any practical implementation of Truncated Versions of these distributions.

Gradients were adapted from the STAN implementation, and require the use of two extra "pseudo aesara ops" that simply wrap the equivalent python functions. This does not allow for further graph optimizations, but I doubt aesara could do much with the previous scan implementation either. Let me know if you think there is a better solution.

Depending on what your PR does, here are a few things you might want to address in the description:
+ [x] what are the (breaking) changes that this PR makes? dist_math::incomplete_beta was deprecated in favor of the new betainc op
+ [x] important background, or details about the implementation
+ [x] are the changes—especially new features—covered by tests and docstrings?
+ [x] [linting/style checks have been run](https://github.com/pymc-devs/pymc3/wiki/PyMC3-Python-Code-Style)
+ [x] [consider adding/updating relevant example notebooks](https://github.com/pymc-devs/pymc-examples)
+ [x] right before it's ready to merge, mention the PR in the RELEASE-NOTES.md

Closes #4420

brandonwillard committed:
Initial refactoring for RandomVariable use in Model, step methods, and basic dists

These changes can be summarized as follows:
- `Model` objects now track fully functional Theano graphs that represent all
relationships between random and "deterministic" variables.  These graphs are
called these "sample-space" graphs.  `Model.unobserved_RVs`, `Model.basic_RVs`,
`Model.free_RVs`, and `Model.observed_RVs` contain these
graphs (i.e. `TensorVariable`s), which are generated by `RandomVariable` `Op`s.
- For each random variable, there is now a corresponding "measure-space"
variable (i.e. a `TensorVariable` that corresponds to said variable in a
log-likelihood graph).  These variables are available as `rv_var.tag.value_var`,
for each random variable `rv_var`, or via `Model.vars`.
- Log-likelihood (i.e. measure-space) graphs are now created for individual
random variables by way of the generic functions `logpt`, `logcdf`,
`logp_nojac`, and `logpt_sum` in `pymc3.distributions`.
- Numerous uses of concrete shape information stemming from `Model`
objects (e.g. `Model.size`) have been removed/refactored.
- Use of `FreeRV`, `ObservedRV`, `MultiObservedRV`, and `TransformedRV` has been
deprecated.  The information previously stored in these classes is now tracked
using `TensorVariable.tag`, and log-likelihoods are generated using the
aforementioned `log*` generic functions.

brandonwillard committed:
Temporarily disable CI tests

brandonwillard committed:
Remove shape dependencies from DictToArrayBijection

This commit changes `DictToArrayBijection` so that it returns a `RaveledVars`
datatype that contains the original raveled and concatenated vector along with
the information needed to revert it back to dictionay/variables form.

Simply put, the variables-to-single-vector mapping steps have been pushed away
from the model object and its symbolic terms and closer to the (sampling)
processes that produce and work with `ndarray` values for said terms.  In doing
so, we can operate under fewer unnecessarily strong assumptions (e.g. that the
shapes of each term are static and equal to the initial test points), and let
the sampling processes that require vector-only steps deal with any changes in
the mappings.

brandonwillard committed:
Rename Model.ndim to Model.size

This value was not representative of its name.

brandonwillard committed:
Update competence methods to work with RandomVariables

brandonwillard committed:
Implement naive RandomVariable-based posterior predictive sampling

The approach currently being used is rather inefficient.  Instead, we should
change the `size` parameters for `RandomVariable` terms in the sample-space
graph(s) so that they match arrays of the inputs in the trace and the desired
number of output samples.  This would allow the compiled graph to vectorize
operations (when it can) and sample variables more efficiently in large batches.

brandonwillard committed:
Remove newly deprecated classes and functions

Classes and functions removed:
- PyMC3Variable
- ObservedRV
- FreeRV
- MultiObservedRV
- TransformedRV
- ArrayOrdering
- VarMap
- DataMap
- _DrawValuesContext
- _DrawValuesContextBlocker
- is_fast_drawable
- _compile_theano_function
- vectorize_theano_function
- get_vectorize_signature
- _draw_value
- draw_values
- generate_samples
- fast_sample_posterior_predictive

Modules removed:
- pymc3.distributions.posterior_predictive
- pymc3.tests.test_random

fonnesbeck committed:
Removed redundant bound in Wald distribution

brandonwillard committed:
Refactor tests for compatibility with logp dispatch and RandomVariables

OriolAbril committed:
Port InferenceData conversion code

michaelosthege committed:
Apply easy fixes to get tests to pass or xfail

michaelosthege committed:
Allow ignoring tests files, but print a warning about it

michaelosthege committed:
Make XPASSing tests fail, so we'll know when something is fixed

Closes #4516

michaelosthege committed:
Ignore tests that are completely broken, but run the others

ricardoV94 committed:
Add tests for gradient of betainc

OriolAbril commented:
not sure if this should be merged into master or into v4 :thinking: cc @brandonwillard @twiecki @michaelosthege

brandonwillard commented:
It shouldn't be difficult to rebase `v4`, so we could merge something like this.  I think the problem would have more to do with our release plan (e.g. are we going to make version 3 releases with the `theano` -> `aesara` change?)

ricardoV94 commented:
Strange failure on 32-bit run, for the logcdf of the beta.

```python
E               AssertionError: 
E               Arrays are not almost equal to 3 decimals
E               {'alpha': array(100., dtype=float32), 'beta': array(0.1, dtype=float32), 'value': array(0.001, dtype=float32)}
E               x and y -inf location mismatch:
E                x: array(-inf, dtype=float32)
E                y: array(-697.172)
```

I don't understand why the aesara op would underflow but not the scipy one. It seems scipy is basically calling betainc and then taking the logarithm, which is exactly the same thing we are doing: 

https://github.com/pymc-devs/pymc3/blob/53b642e4124feb0092ef5492685f44420f80cf3b/pymc3/distributions/continuous.py#L1319

https://github.com/scipy/scipy/blob/master/scipy/stats/_continuous_distns.py#L610-L611

Afaik the scipy should also underflow to 0 and then -inf in a float-32 environment:

```python
np.log(st.beta(100.0, 0.1).cdf(0.001).astype('float32'))
<ipython-input-32-b9435d6250d7>:1: RuntimeWarning: divide by zero encountered in log
  np.log(st.beta(100.0, 0.1).cdf(0.001).astype('float32'))
Out[32]: -inf
```

michaelosthege commented:
@ricardoV94 this error could indicate that `y` ran with float64:
```
E                x: array(-inf, dtype=float32)
E                y: array(-697.172)
```

michaelosthege commented on RELEASE-NOTES.md:
```diff
@@ -3,10 +3,12 @@
 ## PyMC3 vNext (TBD)
 ### Breaking Changes
 - ⚠ Theano-PyMC has been replaced with Aesara, so all external references to `theano`, `tt`, and `pymc3.theanof` need to be replaced with `aesara`, `aet`, and `pymc3.aesaraf` (see [4471](https://github.com/pymc-devs/pymc3/pull/4471)).
+- The `incomplete_beta` function in pymc3.distributions.dist_math was replaced by an equivalent fast vectorized `betainc` aesara op (see [4519](https://github.com/pymc-devs/pymc3/pull/4519)).
```
```suggestion
- The `incomplete_beta` function in pymc3.distributions.dist_math was replaced by an equivalent fast vectorized `betainc` Aesara `Op` (see [4519](https://github.com/pymc-devs/pymc3/pull/4519)).
```

brandonwillard committed:
Reinstate log-likelihood transforms

ricardoV94 commented:
> @ricardoV94 this error could indicate that `y` ran with float64:
> ```
> E                x: array(-inf, dtype=float32)
> E                y: array(-697.172)
> ```

I was probably unclear. The error appeared in the 32bit GH run: https://github.com/pymc-devs/pymc3/runs/2075254035?check_suite_focus=true

AFAIK both should have underflowed.

ricardoV94 committed:
Strip observed in prior predictive

ricardoV94 commented:
In terms of V3 / V4, I think it's fine if this only appears in the V4 branch. I can toggle this as a draft until we are ready for it.

twiecki commented on RELEASE-NOTES.md:
```diff
@@ -3,10 +3,12 @@
 ## PyMC3 vNext (TBD)
 ### Breaking Changes
 - ⚠ Theano-PyMC has been replaced with Aesara, so all external references to `theano`, `tt`, and `pymc3.theanof` need to be replaced with `aesara`, `aet`, and `pymc3.aesaraf` (see [4471](https://github.com/pymc-devs/pymc3/pull/4471)).
+- The `incomplete_beta` function in pymc3.distributions.dist_math was replaced by an equivalent fast vectorized `betainc` Aesara `Op` (see [4519](https://github.com/pymc-devs/pymc3/pull/4519)).
```
```suggestion
- The `incomplete_beta` function in `pymc3.distributions.dist_math` was replaced by an equivalent fast vectorized `betainc` Aesara `Op` (see [4519](https://github.com/pymc-devs/pymc3/pull/4519)).
```

ricardoV94 committed:
Fix values in `test_ignores_observed`

brandonwillard committed:
Use name from Variable.name in pymc3.util.get_var_name

brandonwillard committed:
Add fallback for missing value_var in BaseTrace

ricardoV94 commented:
Waiting a bit on this one following some discussion on the STAN side to make sure these gradient approximations are good enough: https://github.com/stan-dev/math/issues/2417

brandonwillard committed:
Remove remaining v3 sampling code

brandonwillard committed:
Change logp_transform argument to the entire random variable

brandonwillard committed:
Remove logpt transformed option

brandonwillard committed:
Implement transform for Dirichlet

brandonwillard committed:
Always use the value var to initially build the log-likelihood

brandonwillard committed:
Register GammaRV to PositiveContinuous class

brandonwillard committed:
Add an option for negative support assertions in Normal and Gamma classes

brandonwillard committed:
Make logp testing work with transformed values

brandonwillard committed:
Disable asserts during logp invalid range tests

brandonwillard committed:
Update xfails in pymc3.tests.test_distributions

brandonwillard committed:
Fix Categorical logp implementation

brandonwillard committed:
Use the value var's values directly in logpt

brandonwillard committed:
Do not assume values are Applys in pymc3.tests.test_distributions

kc611 committed:
Refactored distributions in pymc.distributions.continuous

brandonwillard committed:
Simplify the new Distribution interface and convert a few more

brandonwillard committed:
Use xfail mark in pymc3.tests.test_distributions_timeseries

brandonwillard committed:
Create class after setting the class dictionary

ricardoV94 committed:
Fix HalfCauchy/HalfCauchyRV parameterization

brandonwillard committed:
Make transform objects stateless

ricardoV94 committed:
Fix shape logic in test_distributions_random and skip tests covered by Aesara

ricardoV94 committed:
Re-enable test_distributions_random

ricardoV94 committed:
Refactor Beta to use custom rng_fn clipped_beta_rv

ricardoV94 committed:
Fix HalfNormal/HalfNormalRV parameterization

brandonwillard committed:
Add tests for two important open logpt and Model issues

brandonwillard committed:
Add non_sequences to uses of Scan Op

This make `aesara.graph.basic.clone_replace` work correctly when `Scan`s are
included in a graph.

brandonwillard committed:
Replace Observed Op with tag.observations

brandonwillard committed:
Add missing imports to pymc3.step_methods.gibbs

brandonwillard committed:
Comment out unused moments

brandonwillard committed:
Make logpt work correctly for nested models and transforms

brandonwillard committed:
Disable use of Arviz in pymc3.tests.test_data_container

brandonwillard committed:
Set model seed correctly in pymc3.tests.test_ndarray_backend

brandonwillard committed:
Prevent Model from turning on test value computations

brandonwillard committed:
Make Metropolis, Slice, PGBART, MetropolisMLDA use point values

brandonwillard committed:
Re-enable disabled tests

brandonwillard committed:
Use value vars to determine steps

brandonwillard committed:
Use size instead of shape in pymc3.tests.sampler_fixtures

brandonwillard committed:
Set model-level RandomVariable seeds during sampling

brandonwillard committed:
Check shapes by evaluating graph with start values

brandonwillard committed:
Fix logpt so that transforms are always applied, when enabled

brandonwillard committed:
Set default transform for Dirichlet

brandonwillard committed:
Normalize Multinomial argument

brandonwillard committed:
Use no_transform_object in Distribution.__new__

brandonwillard committed:
Fix Interval.jacobian_det

brandonwillard committed:
Fix Stickbreaking scalar condition

brandonwillard committed:
Make Model.test_point generation transform existing RV test value

brandonwillard committed:
Add transformed value variables to Model.named_vars

brandonwillard committed:
Remove DeterministicWrapper from Deterministic

brandonwillard committed:
Make sure sample_posterior_predictive doesn't use trace values for sampled vars

brandonwillard committed:
Set seed after loading trace in TestSaveLoad

brandonwillard committed:
Make v4 compatibility changes to pymc3.tests.test_sampling

brandonwillard committed:
Make pymc3.tests.test_transforms work with None RV variables

brandonwillard committed:
Make find_MAP work with RaveledVars

brandonwillard committed:
Make sure start values are NumPy arrays

brandonwillard committed:
Raise NotImplementedError in Group.__init__

brandonwillard committed:
Add type hints to astep methods

brandonwillard committed:
Use untransformed samples and xfail Arviz tests in BaseSampler

brandonwillard committed:
Make sure forward transformed input is a TensorVariable in TestMatchesScipy

brandonwillard committed:
Adjust flaky last-digit numerical requirements in TestMatchesScipy

brandonwillard committed:
Fix MvNormal quaddist_matrix parameter order

brandonwillard committed:
Enable MvNormal tests in test_distributions

brandonwillard committed:
Factor out parameter pre-processing in TestMatchesScipy

brandonwillard committed:
Apply recent xfail updates from master branch

brandonwillard committed:
Prevent dtype conversion in Aesara during testing to avoid a bug

This change avoids the bug fixed in
https://github.com/pymc-devs/aesara/pull/348.

brandonwillard committed:
Fix NegativeBinomial parameterization and enable its tests

brandonwillard committed:
Prevent SciPy error by using float64 point in test_dirichlet_with_batch_shapes

brandonwillard committed:
Disable dims, default_dims, and index_origin options until arviz > v0.11.2

brandonwillard committed:
Create extract_obs_data function

brandonwillard committed:
Re-enable Arviz tests in pymc3.tests.test_sampling

brandonwillard committed:
Do not use shared variables as inputs during prior/posterior sampling

OriolAbril committed:
add workaround for data groups until next arviz release

OriolAbril committed:
activate arviz compat tests

brandonwillard committed:
Remove redundant coords test

brandonwillard committed:
Re-enable Arviz parts of sampler fixtures

brandonwillard committed:
Re-enable v4 xfails in pymc3.distributions.dist_math

brandonwillard committed:
Remove deprecated default values tests

brandonwillard committed:
Re-enable pymc3.tests.test_modelcontext

brandonwillard committed:
Fix extra_vars in call to ValueGradFunction from Model

brandonwillard committed:
Re-enable some tests in pymc3.tests.test_model

brandonwillard committed:
Remove Model.size

brandonwillard committed:
Always set model.default_rng in pymc3.sampling.sample

brandonwillard committed:
Change shape to size in pymc3.tests.test_step

brandonwillard committed:
Mark brittle TestStepMethods.test_sample_exact test as xfail

brandonwillard committed:
Remove incorrect size for multivariate distributions in pymc3.tests.models

brandonwillard committed:
Change shape to size in pymc3.tests.test_aesaraf

brandonwillard committed:
Make minor v4 updates to test_car_logp

brandonwillard committed:
Re-enable more sampler-based tests and selectively xfail others

brandonwillard committed:
Update xfail reason in pymc3.tests.test_examples

brandonwillard committed:
Set a seed for a flaky pymc3.tests.test_sampling test

brandonwillard committed:
Rename Model.vars to Model.value_vars

brandonwillard committed:
Reintroduce the old Model.vars with a deprecation warning

brandonwillard committed:
Add value/rv variable maps to Model

brandonwillard committed:
Change no_transform_object sentinel to UNSET in pymc3.util

brandonwillard committed:
Move logp code into pymc3.distributions.logp and graph functions into pymc3.aesaraf

brandonwillard committed:
Remove build_named_node_tree and its helper functions

brandonwillard committed:
Move treelist and treedict to pymc3.util

brandonwillard committed:
Remove pymc3.model._walk_up_rv

brandonwillard committed:
Remove pymc3.model.as_iterargs

brandonwillard committed:
Move pymc3.model.all_continuous to pymc3.sampling

brandonwillard committed:
Move pymc3.model.pandas_to_array to pymc3.aesaraf

brandonwillard committed:
Rename pymc3.tests.test_model_utils to test_glm_utils

brandonwillard committed:
Rename pymc3.tests.test_models_linear to test_glm_linear

brandonwillard committed:
Add xfail to pymc3.tests.test_variational_inference

brandonwillard committed:
Remove remaining uses of Distribution.default

brandonwillard committed:
Remove Distribution.get_test_val and Distribution.getattr_value

brandonwillard committed:
Rename Model.test_point to Model.initial_point

brandonwillard committed:
Rename Model.check_test_point to Model.point_logps

brandonwillard committed:
Make pymc3.util.update_start_vals and check_start_vals Model methods

brandonwillard committed:
Re-enable pymc3.tests.test_util

brandonwillard committed:
Fix issue with single variable input to BlockedStep

brandonwillard committed:
Make GradientSharedStep a subclass of ArrayStepShared

brandonwillard committed:
Make initial_replacements argument in rvs_to_value_vars optional

brandonwillard committed:
Rename rv_ancestors to walk_model and add tests

brandonwillard committed:
Merge and improve apply_transforms and rvs_to_value_vars

brandonwillard committed:
Fix a NumPy non-tuple indexing deprecation warning in test_aesaraf

brandonwillard committed:
Convert and transform potentials in the log-likelihood

brandonwillard committed:
Finish refactoring BlockedStep.competence implementations

brandonwillard committed:
Fix RaveledVars and size-related issues in Metropolis and MLDA samplers

brandonwillard committed:
Seed flaky TestSamplePPC.test_deterministic_of_observed_modified_interface

brandonwillard committed:
Re-enable step method tests in pymc3.tests.test_step

brandonwillard committed:
Add a test module for pymc3.distributions.logp

brandonwillard committed:
Implement log-likelihoods for *IncSubtensor* Ops

fonnesbeck committed:
Converted Gumbel distribution to v4

fonnesbeck committed:
Ran pre-commit scripts

brandonwillard committed:
Use the Subtensor-ed result as the observed random variable

brandonwillard committed:
Remove unnecessary module pymc3.tests.test_model_helpers

brandonwillard committed:
Update Arviz conversion for new missing values approach

brandonwillard committed:
Update tests for new missing values approach

brandonwillard committed:
Allow empty RaveledVars results in DictToArrayBijection.map

fonnesbeck committed:
Converted Pareto distribution to v4

brandonwillard committed:
Allow dict of value vars in logp signatures and implement Subtensor logp

ricardoV94 committed:
Fix exponential and gamma logp / random link (#4576)

brandonwillard committed:
Add missing values as unobserved random variables and estimate them during MCMC

brandonwillard committed:
Add a starting point option to DictToArrayBijection.rmap

brandonwillard committed:
Make sure shared variables have correct broadcasting in ValueGradFunction

brandonwillard committed:
Fix RaveledVars usage in BinaryGibbsMetropolis.astep

brandonwillard committed:
Use ABC abstract methods in step method interface

brandonwillard committed:
Format find_MAP docstring

brandonwillard committed:
Enable Arviz tests

michaelosthege committed:
Compute the size of scalars for size expansions in pymc3.aesaraf.change_rv_size

larryshamalama committed:
Fixed grammar typos: "a aesara" to "an Aesara" (take 3) (#4624)

* Fixed Aesara related spelling in documentation

ckrapu committed:
Fix sequential backend (#4584)

* Fixing error with sequential chain sampling for NDArray backend (closes #4568)

michaelosthege committed:
Remove nonexistent attribute from __all__ list (#4627)

ricardoV94 committed:
Fix Uniform logp regression from #4541

brandonwillard committed:
Use optimized graph for JAX conversion

ricardoV94 committed:
Refactor DiscreteWeibull (#4615)

brandonwillard committed:
Create a NumPyro sampler Op for better JAX backend integration

ricardoV94 committed:
Add unobserved_value_vars property to Model

Fixes pm.Deterministics still using rvs during pm.Sample()
Fixes find_MAP() failing when pm.Deterministics were present in model

ricardoV94 committed:
Add auto_deterministics list to Model

Ensures that when missing variables are present in the model, the automatic deterministic (x_observed + x_missing) only appears in predictive sampling and not normal sampling.
Fixes `x` missing from prior_predictive when missing values were present (only `x_missing` was present)

michaelosthege committed:
Allow unnamed (None) dims and undefined (None) coord values

Also refactor into properties to add docstrings and type annotations.
And no longer allow InferenceData conversion without a Model on stack.

Co-authored-by: Oriol Abril Pla <oriol.abril.pla@gmail.com>

michaelosthege committed:
Allow parametrization through either shape, dims or size

Always treats `size` as being in addition to dimensions implied by RV parameters.

All resizing beyond parameter-implied dimensionality is done from:
- `shape` or `size` in `Distribution.dist()`
- `dims` or `observed` in `Distribution.__new__`
and only in those two places.

Closes #4552.

michaelosthege committed:
Automatically add SpecifyShape Op when full-length shape is given

ricardoV94 committed:
Fix missing fstrings

Spaak committed:
remove obsolete multiplication with 1.0 in get_sigma_tau

MarcoGorelli committed:
fixup stacklevel

matteo-pallini committed:
V4 update test framework for distributions random method (#4608)

* Update tests following distributions refactoring

The distributions refactoring moves the random variable sampling to
aesara. This relies on numpy and scipy random variables implementation.
So, now the only thing we care about testing is that the parametrization
on the PyMC side is sendible given the one on the Aesara side
(effectively the numpy/scipy one)

More details can be found on issue #4554
https://github.com/pymc-devs/pymc3/issues/4554

* Change tests for more refactored distributions.

More details can be found on issue #4554
https://github.com/pymc-devs/pymc3/issues/4554

* Change tests for refactored distributions

More details can be found on issue #4554
https://github.com/pymc-devs/pymc3/issues/4554

* Remove tests for random variable samples shape and size

Most of the random variable logic has been moved to aesara, as well as
most of the relative tests. More details can be found on issue #4554

* Fix test for half cauchy, renmae mv normal tests and add test for
Bernoulli

* Add test checking PyMC samples match the aesara ones

Also mark test_categorical as expected to fail due to bug on aesara
side. The bug is going to be fixed with 2.0.5 release, so we need to
bump the version for categorical and the test to pass.

* Move Aesara to 2.0.5 to include Gumbel distribution

* Enamble exponential and gamma tests following bug-fix

* Enable categorical test following aesara version bump to 2.0.5 and relative bug-fix

* Few small cosmetic changes:
- replace list of tuples with dict
- rename 1 method
- move pymc_dist as first argument in function call
- replace list(params) with params.copy()

* Remove redundant tests

* Further refactoring

The refactoring should make it possible testing both the distribution
parametrization and sampled values according to need, as well as any
other future test. More details on PR #4608

* Add size tests to new rv testing framework

* Add tests for multivariate and for univariate multi-parameters

* remove test already covered in aesara

* fix few names

* Remove "distribution" from test class names

* Add discrete Weibull, improve Beta and some minor refactoring

* Fix typos in checks naming and add sanity check

Co-authored-by: Ricardo <ricardo.vieira1994@gmail.com>

larryshamalama committed:
Fixing np.bool Depreciation warnings: 3 files (#4673)

* :turtle: Fixed documentation

* :turtle: Minor fix on top of grammar fixed in docs

* Weibull RandomVariable refactoring: an example

* Revert "Weibull RandomVariable refactoring: an example"

This reverts commit 685a06554195eeb73b07fceea90fb783187990cb.

* Fixing np.bool depreciation warnings: 3 files

* Changed np.bool_ to bool (+ fixed other places)

ricardoV94 committed:
Refactor DiscreteUniform

ricardoV94 committed:
Refactor Constant

ricardoV94 committed:
Refactor OrderedLogistic

ricardoV94 committed:
Refactor OrderedProbit

ricardoV94 committed:
Add missing discrete distributions to API rst

ricardoV94 committed:
Refactor ZeroInflatedPoisson

ricardoV94 committed:
Refactor ZeroInflatedBinomial

ricardoV94 committed:
Refactor ZeroInflatedNegativeBinomial

ricardoV94 committed:
Update several test xfails

farhanreynaldo committed:
Remove GLM submodule in v4 (#4682)

* remove GLM submodule and its test

* add glm reference to Bambi

* add GLM removal to release notes

Co-authored-by: Farhan Reynaldo <farhanreynaldo@gmail.com>

ricardoV94 committed:
Refactor several distributions (#4640)

* Refactor several distributions

* Fix continuous bounded default transform

* Add 32bit xfail to Weibull logp

* Add TODO reminder for Weibull

* Refactor random tests

* Remove tests covered by Aesara

* Refactor BetaBinomial

* Add skipif for betabinom depending on scipy version

* Rename transform_args to bound_args_indices

ricardoV94 committed:
Update dependency to aesara 2.0.8, and necessary fixes

ricardoV94 committed:
Replace custom incomplete_beta function with scipy betainc and respective derivatives

ricardoV94 committed:
Use _logp and _logcdf dispatcher in ZeroInflated* methods

ricardoV94 committed:
Fix `check_logcdf` test regression

brandonwillard committed:
Revert "Automatically add SpecifyShape Op when full-length shape is given"

This reverts commit c99f15c28f5fd6635f38e0d130f49d410501bc9b.

brandonwillard committed:
Revert "Allow parametrization through either shape, dims or size"

This reverts commit ed29203f447c7b6d1687e31a3eda3da000c349df.

farhanreynaldo committed:
refactor pareto and laplace (#4691)

* refactor pareto
* refactor laplace
* Reintroduce `Pareto` default transform

Co-authored-by: Farhan Reynaldo <farhanreynaldo@gmail.com>
Co-authored-by: Ricardo <ricardo.vieira1994@gmail.com>

ricardoV94 committed:
Fix BinaryMetropolis astep

ricardoV94 committed:
Fix _check_start_shape

ricardoV94 committed:
Update failing tests

themrzmaster committed:
Refactor Student T Distribution (#4694)

* feat: adapt student t
* Change default parameterization in terms of sigma and tweak tests

Co-authored-by: Ricardo <ricardo.vieira1994@gmail.com>

brandonwillard committed:
Remove duplicate pandas_to_array call in Model.register_rv

brandonwillard committed:
Make sure new size values are int64

Closes #4652.

brandonwillard committed:
Fix mistaken use of change_rv_size with new_size None

michaelosthege committed:
Edit misleading comment to be more accurate

farhanreynaldo committed:
refactor Moyal distribution (#4704)

* refactor Moyal distribution

* fixed ndim_params on moyal

Co-authored-by: Farhan Reynaldo <farhanreynaldo@gmail.com>

farhanreynaldo committed:
refactor kumaraswamy (#4706)

* refactor kumaraswamy
* add logcdf method

Co-authored-by: Farhan Reynaldo <farhanreynaldo@gmail.com>
Co-authored-by: Ricardo <ricardo.vieira1994@gmail.com>

farhanreynaldo committed:
Refactor Rice and Skew Normal distribution (#4705)

* refactor Skewnormal and Rice distribution
* add test for rice b and skewnorm tau params
* change default parameter to b
* Add float32 xfail to `test_rice`

Co-authored-by: Farhan Reynaldo <farhanreynaldo@gmail.com>
Co-authored-by: Ricardo <ricardo.vieira1994@gmail.com>

themrzmaster committed:
Refactor LogitNormal (#4703)

MarcoGorelli committed:
let pandas_to_array take pandas Index

brandonwillard committed:
Update Aesara requirement to 2.0.9

brandonwillard committed:
Use aesara.tensor.atleast_1d in pymc3.aesaraf.change_rv_size

ricardoV94 committed:
Add testval to flaky test (#4707)

brandonwillard committed:
Incrementally update the RNG state in Model

ricardoV94 committed:
Refactor Flat and HalfFlat distributions (#4723)

* Refactor Flat and HalfFlat distributions

* Re-enable Gumbel logp test

* Remove redundant test

brandonwillard committed:
Convert RandomVariables to in-place during graph optimization

brandonwillard committed:
Introduce Model.initial_values and deprecate testval in favor of initval

brandonwillard committed:
Replace uses of testval with initval

brandonwillard committed:
Remove unnecessary gammaln and psi from pymc3.distributions.special

ricardoV94 committed:
Revert reduced test n_samples due to speed issues

ricardoV94 committed:
Speedup check_logcdf test

ricardoV94 committed:
Speedup check_selfconsistency_discrete_logcdf test

ricardoV94 committed:
Float32 xfail on Beta and StudentT logcdf tests

ricardoV94 committed:
Add workaround for HyperGeometric logcdf failure

ricardoV94 committed:
Remove `gammainc(c)` safeguards in `logcdf` methods of `Gamma` and `InverseGamma`

Closes #4467
